[preview] Openpulse schedule

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Fake PR to clarify the diff


### Details and comments

itoko committed:
wip: make schedule immutable

itoko committed:
wip: make schedule immutable

itoko committed:
make schedule immutable

itoko committed:
make schedule immutable

itoko committed:
fix a bug (remove an accidental comment out)

itoko committed:
fix a bug in copying of a Schedule

itoko committed:
fix a bug in copying of a Schedule

itoko committed:
Merge branch 'commands-rework' into schedule-rework-oop

Commit:
Merge branch 'master' of github.com:Qiskit/qiskit-terra into schedule-rework-oop

Commit:
Merge branch 'schedule-rework-oop' into openpulse-schedule

nkanazawa1989 commented:
Do not merge this

taalexander commented on qiskit/pulse/commands/acquire.py:
```diff
@@ -80,8 +80,14 @@ def __call__(self,
                  reg_slots: Union[RegisterSlot, List[RegisterSlot]] = None) -> 'AcquireInstruction':
         return AcquireInstruction(self, qubits, mem_slots, reg_slots)
 
+    def __rshift__(self, args) -> 'AcquireInstruction':
```
I do not think instructions should be initialized in this way. If we do, it should not be right-shift as this should be reserved for time shifting of schedules.

itoko replied:
I know. I'll remove this.

itoko replied:
done

taalexander commented on qiskit/pulse/commands/persistent_value.py:
```diff
@@ -56,8 +56,11 @@ def __repr__(self):
+class PersistentValueInstruction(PrimitiveInstruction):
     """Pulse to keep persistent value. """
 
     def __init__(self, command: PersistentValue, channel: OutputChannel):
```
This `init` sequence is repeated in every implemented instruction. This should be rewritten as a call to the superclass initializer with the proper arguments.

taalexander commented on qiskit/pulse/common/command_schedule.py:
```diff
@@ -0,0 +1,78 @@
+logger = logging.getLogger(__name__)
+
+
+class PrimitiveInstruction(Instruction):
```
I don't see what the difference is between an `Instruction` and a `PrimitiveInstruction` there are starting to be too many slightly different classes and it is confusing from an API perspective.

taalexander commented on qiskit/pulse/common/command_schedule.py:
```diff
@@ -0,0 +1,78 @@
+        return []
+
+    @property
+    def begin_time(self) -> int:
```
t0 or start_time, I think t0 is less verbose.

itoko replied:
I used `begin_time` because pylint don't like `t0`...

taalexander replied:
This is fine we can take one of two approaches.
1. Disable warnings and use `t0`
2. Use `start_time`

I do think that technically `begin_time` and `end_time` are technically correct. However, within American English when people talk about timers, they will `start` and `stop` a timer. This is why I would opt for `start_time` and `stop_time`. This is really a fault of the english language.

taalexander commented on qiskit/pulse/common/command_schedule.py:
```diff
@@ -0,0 +1,78 @@
+        return self._block
+
+    @property
+    def end_time(self) -> int:
```
tf or stop_time.

taalexander commented on test/python/pulse/test_schedule.py:
```diff
@@ -0,0 +1,108 @@
+from qiskit.test import QiskitTestCase
+
+
+class TestSchedule(QiskitTestCase):
```
We will require more individual unittests of all classes/methods to get this merged into master. Another reason why decreasing the complexity of the API will make our lives easier.

taalexander commented on qiskit/pulse/schedule/pulse_schedule.py:
```diff
@@ -0,0 +1,205 @@
+    def name(self) -> str:
+        return self._name
+
+    def inserted(self, begin_time: int, instruction: Instruction):
```
`insert`.

itoko replied:
I intended `return a new object` by `ed` like `sorted()` for list. Anyway I changed back to `insert` since can discuss renames later.

taalexander commented on qiskit/pulse/schedule/pulse_schedule.py:
```diff
@@ -0,0 +1,205 @@
+            logger.warning("Fail to insert %s at %s due to timing overlap", inst, begin_time)
+            raise PulseError("Fail to insert %s at %s due to overlap" % (str(inst), begin_time))
+
+    def appended(self, instruction: Instruction):
```
append

taalexander reviewed (changes requested):
## Summary
I think this is on the right track, especially with its combination of composition and inheritance. However, I think the current implementation is confusing with much redundancy from a result of trying to be too modular and in certain places has taken the composition pattern too far. It is very difficult to keep in my head, and this is from the perspective of someone who is familiar. It would be very hard for our users to grasp.

Additionally, many more tests of the individual components will be required to merge into Terra master.  I believe the changes I suggest will reduce this amount by reducing the complexity of the API.

## What we need
- `union` method for classes that implement a `ScheduleInterface`.
- `Schedule`
     - Contains children that implement `ScheduleInterface` inserted at times. 
     - Derives timing information from the union of child schedules.
     - Is based at time 0.
- `CommandSchedule`
     - Form the leaf nodes of a `Schedule`.
     - Has a `Command`
     - Derives duration from its `Command`.
- `Command`
     - Has duration
     - Specific command parameters.
## What we have
- `ScheduleNode`
     - Questions:
          - Is an interface for what? Something with children? It is only inherited by `TimedInstruction`
- `TimedInstruction`
      - This is essentially a `CommandSchedule` why do we need all of the additional classes?
- `Instruction`
      - Is also just an interface?
- `CommandSchedule`
     -  Once again `TimedInstruction` pops up but it is unclear what role it fulfills.
- `SubSchedule`
     - Inherits:
          - `TimedInstruction`
     - Why is this not just a new Schedule?
- `Schedule`
      - Inherits from both `TimedInstruction` and `TimedInstruction`, why not just `ScheduleInterface`?
- `PrimitiveInstruction`
     - How is this different then a `CommandSchedule`?


## Suggestions for this branch before submitting
- Combine `ScheduleNode` and `TimedInstruction` in `ScheduleInterface`.
- Rename interface classes so it is clear they are interfaces, ie., `ScheduleInterface`
- Schedules should simply store the time a child `Schedule` starts at. No need for another object which obfuscates the object it is composed of.
     - This removes need for `SubSchedule` class.
- Calling a `Command` returns a `CommandSchedule` which has a `Command` instance, as far as I can see there is no need for the `PrimitiveInstruction` class, it is confusing. If necessary could create individual `CommandSchedule`s for each type of `Command`  but I don't think necessary.

itoko commented:
@taalexander Thank you for helpful review. I introduced`TimedInstruction` trying to reuse Schedule object as much as possible but, reading your review, I found the possibility to make Schedule reusable without introducing `TimedInstruction` interface. I'll try to do so, and it will remove SubSchedule as you suggested.

itoko committed:
refactor (simplify) schedule module

itoko committed:
refactor (simplify) schedule module

itoko committed:
Merge pull request #38 from nkanazawa1989/schedule-rework

Refactor (simplify) schedule module

itoko commented:
@taalexander I think I succeeded to refactor schedule module along with your review comment. 

By 3152312, I removed unnecessary interfaces/classes: `ScheduleNode`, `TimedInstruction`, `SubSchedule` and `PrimitiveInstruction`.
Now we have simplified interfaces/classes:
- Instruction (~= CommandSchedule): Leaf node of schedule tree.
- Schedule
- ScheduleComponent (~= ScheduleInterface): Common interface of Instruction and Schedule
Note that Instruction changed its meaning before and after this PR.

I think we still have naming issues, e.g. t0/begin_time or Instruction/CommandSchedule, but I think this PR now become almost ready to go.

itoko committed:
Merge branch 'openpulse-commands-rework' into openpulse-schedule

itoko committed:
add command property to instruction for assemler

itoko committed:
add command property to instruction for assemler

nkanazawa1989 commented on qiskit/pulse/common/__init__.py:
```diff
@@ -1,8 +0,0 @@
-# -*- coding: utf-8 -*-
```
You need to keep copyright.

itoko replied:
Nice catch. Done.

itoko committed:
restore accidentally deleted file

itoko committed:
add a corner case test

itoko committed:
add a corner case test

taalexander commented on qiskit/pulse/commands/instruction.py:
```diff
@@ -0,0 +1,65 @@
+    def __repr__(self):
+        return "%4d: %s" % (self._begin_time, self._command)
+
+    def __lshift__(self, shift: int) -> ScheduleComponent:
```
Could we add __rshift__ too which would just be `self.shifted(-shift)`?

itoko replied:
We could, but I think it's optional, users can do all they want without rshift. Shall we discuss this later?

taalexander replied:
I think we should either add it or remove __lshift__ in this PR. It will be confusing if one is available and the other is not.

taalexander commented on qiskit/pulse/schedule/pulse_schedule.py:
```diff
@@ -5,201 +5,158 @@
+            PulseError: when an invalid schedule is specified or failed to insert
         """
-        if inst == self:
+        if schedule == self:
```
I believe this requirement should be easy to relax but we can do this in future PR.

itoko replied:
I think you are right. It's already safe unless the private `_insert()` is directly called. Just doing `__insert()`? Let's discuss later PR.

taalexander reviewed (changes requested):
This looks excellent @itoko! Exactly what I was envisioning and seems to be much easier to understand. 

Just a couple recommended naming changes.
1. Can we Make `begin_time` `start_time` it just sounds better in American english, whereas `begin_time` does not feel right. I have discussed this with @dcmckay. Same goes for `end_time` vs. `stop_time`.
2. `TimeslotOccupancy` as a name also feels weird. Could it be something like`ScheduleTimeslot` and `ChannelTimeslot` for `TimeSlot`? Any other suggestions would be appreciated.

nkanazawa1989 committed:
OpenPulse frontend - commands module rework (#2104)

* OpenPulse Command Group:

commit 3a4034d0150d18e336106071fa52abaa369619e7
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 17:25:30 2019 +0900

    edit CHANGELOG

commit 5a00c0930486abbdab0714f6efddb169669d06d9
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 15:59:26 2019 +0900

    code cleanup

commit 123b110ae8bcc45fad7f49dd58e7653533381e27
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 15:09:04 2019 +0900

    Update plot method

commit a413aa1b4803958135dbc379b58f99f6f4c1befe
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 13:21:20 2019 +0900

    Add test case

commit 9a3136aebb95c9ccc94284df762c9cb375539ad5
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 02:06:55 2019 +0900

    Add pulse validity check to functional_pulse

commit 6915d9cdfc17f7ce86dae972a43346effcebac9e
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 01:56:25 2019 +0900

    Add __init__ to command group and change module names

commit dab9b71606a0faee3d244578460e00b36182f218
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 01:43:11 2019 +0900

    Add acquire, frame change, snapshot, persistent_value

commit ddd6cf60517a76ab0b4e9e2764d7b59533d8b9da
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Feb 12 01:29:01 2019 +0900

    Add draw method to SamplePulse

commit 07b0b0ad2728818ad8ce7cbd5eced7b80b2788c3
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Mon Feb 11 11:48:33 2019 +0900

    fix module name

commit 8ad8dbdcf565cc494a4809c84cdde24f1336c9c8
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Mon Feb 11 10:40:03 2019 +0900

    move functional pulse to command module

commit 2e8a0c5bd55ec0725cb80787aaeb8f2a7f8fad73
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 14:45:50 2019 +0900

    correct module name

commit ab1699eb7efcec25d8987add792757b2a0224225
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 14:31:32 2019 +0900

    reformat code

commit 07e13f08d495897cc710786ef22a3d44bbc8e76f
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 13:27:44 2019 +0900

    add test cases for functional pulse

commit f72f666b9957c0263c2c829de465ed7ec7aa43f3
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 12:57:50 2019 +0900

    add savefig option in plot()

commit a6b34ac2d064fd8fc2e5e58ba187aebcb38cb80c
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 12:07:14 2019 +0900

    allow integer value input to tolist()

commit 6367e3ebfc2612594b5432c247442a944194f7f3
Merge: 6d5f824f ccccd07e
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 06:44:17 2019 +0900

    Merge remote-tracking branch 'origin/pulse_class' into pulse_class

commit 6d5f824faa9846a6e346536efd5f190b99a2a86b
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Wed Jan 23 06:42:47 2019 +0900

    remove scipy requirement from setup.py

commit ccccd07ea1fb1102b74a71d5472ee320586ae79f
Merge: f7335059 f239ff9a
Author: Ali Javadi-Abhari <ajavadia@users.noreply.github.com>
Date:   Tue Jan 22 16:39:14 2019 -0500

    Merge branch 'master' into pulse_class

commit f7335059d75c39c1627fb9f0394ba90df476730f
Author: nkanazawa <39517270+nkanazawa1989@users.noreply.github.com>
Date:   Tue Jan 22 14:39:32 2019 +0900

    Update CHANGELOG.rst

commit fb36a13b6033ef84d18570a1bc055156b5996f6a
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Jan 22 13:39:04 2019 +0900

    add legend and include import in plot()

commit e5cf271030556ffe717b1bae2ce8934ce4b72160
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Jan 22 01:45:25 2019 +0900

    optimize import

commit 667e4fda7d8fc10d3d08fcbf186fdfa1ee4e20d8
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Tue Jan 22 01:28:08 2019 +0900

    separate visualization, use callables instead of str parse

commit 07173bd897688809239ea7b28dcf1ceaedca3f01
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 17:17:03 2019 +0900

    param -> params

commit cfdcde40648f5a561b380789741cfb4f6243a55c
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 14:08:12 2019 +0900

    correct copyright

commit 3121c572a22669f235e5c4bc0729284d456ef70f
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 13:26:55 2019 +0900

    reformat code

commit 8668c882d57aa39ef753310d0de21a77f5730a98
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 09:20:47 2019 +0900

    reformat code

commit 472b585ea0378e56c29ed915eabcd2a83511da4b
Author: knzwnao <knzwnao@jp.ibm.com>
Date:   Fri Jan 18 07:22:31 2019 +0900

    add pulse class to manage pulse envelope

* Remove insignificant pylint disables

* update pulse_drawer

* Update qiskit/pulse/commands/_functional_pulse.py

remove abbreviation

Co-Authored-By: nkanazawa1989 <39517270+nkanazawa1989@users.noreply.github.com>

* remove abbreviation

* Reformat docstring and parameter names

* import PulseCommand and SamplePulse in __init__

* remove memory/register slot arguments

* add name to all commands

* add t0 to pulse command

* change import of superclass and remove leading underscores

* minor revisions

* fix xrange when interpolated

* remove timing information from command gruop

* initial commit (channels)

* factoring a bit

* WIP: add pulse_schedule

* WIP: add minimal interfaces

* minimal implementation

* fix typo

* change Schedule a it

* WIP: refactor channels

* fix supported commands

* add lo_frequency as member

* add exceptions for pulse

* add channel factory

* add command library

* WIP execute initial commit

* WIP complete qobj converter flow

* exclude backend embedding from schedule_to_qobj

* Add data structure of PulseCommands to generate Qobj

* add superclass of measurement options

* refactor channels

* fix bug of sample pulse without name

* fix bug of __eq__ in sample pulse

* fix a bug

* refactor add() and add name

* refactor channels

* add exceptions for pulse

* remove pylint error

* fix bugs

* fix a bug in channel name

* expose channel factory

* refactor channel factory

* fix bugs on naming

* fix overlap check bug

* remove pulse_lib validation

* change command library from set to list

* fix a bug

* fix config bug

* add pulses to lib only when it is SamplePulses

* simplify schedule

* lint

* refactor a bit

* add channel register

* fix bugs

* add channel memory

* refactor channel factory

* debug

* refactor schedule

* wip: look for better memory interface

* rename to channel bank

* refactor channel register

* refactor channels

* refactor channel registers

* fix bugs in channels

* add getters

* fix schedule to follow channel refactoring

* improve error message

* add interface to get lo_freqs

* improve logging

* add backend config to channel bank

* fix backend import bug

* remove name from channel register

* fill channel bank constructor

* fix style errors

* explicit import

* remove supported channel definition from subclasses

* remove config from compiler argument

* fix raises in schedule

* change method name

* get lo_frequencies from channel bank

* remove frequency embedding at compiler

* explicit import

* avoid setting useless kernel/discriminator

* add setter for lo

* raise error when schedule fails to add

* rename channel bank->store and change its interface

* rename ChannelRegister to ChannelList

* add PulseSchedule.append()

* add pulse and QASM models to qiskit.qobj

* add PulseQobj and QASMQobj to qiskit.qobj

* add interface of assemble_schedules

* add unit test

* resolve errors

* remove duplicated dict key in unittest

* modify imports

* PulseQobj/QASMQobj inherit from Qobj

* move qobj generation part to another PR

* add device_spec replacing channel_store

* remove channel_list

* add __eq__ to device_spec

* replace FunctionalPulse class by @function decorator

* rename add->insert, PulseSchedule->Schedule

* avoid name to be required

* remove unused import

* update init of qiskit.pulse

* remove unused FunctionalPulse

* update test for decorator

* a few tweaks

* change interface of Schedule.insert()

* remove FunctionalPulse

* fix imports

* fix pulse.__init__

* fix a bug in append()

* remove name from function

* remove name check from unittest

* add mem_slots, reg_slots and remove snapshot_channel

* add custom validator for measurement options

* restore snapshot channel

* remove n_registers==n_qubits assumption

* add instruction, header, config to BaseQobjExperimentSchema

* update class names (remove Base- prefix, and QASM > Qasm)

* remove PulseQobjHeader, PulseQobjExperimentHeader, QasmQobjHeader, QasmQobjExperimentHeader,

* separate models into specific modules

* Add minor changes

* avoid cyclic import

* add minimal tests for channels

* change prefix of register 'r'->'c'

* lint

* add change log

* more linting

* make commands callable

* make channels hashable

* add working example code of Schedule

* add assemble_schedule and execute_schedule

* reorder class defs and add prints

* simplify Schedule

* rename PulseChannel->Channel

* lint (temporal)

* replace dict overwrite by .update

* rename PulseChannel->Channel

* lint

* add getters for qobj creation

* simplify snapshot command

* rename Pulse -> Instruction

* add DeviceSpecification to pulse.__init__

* fix a bug when u_channels==0

* add test for fixing the u_channel==0 case bug completely

* fix #15

* simplify Instruction

* assign unique name for all commands

* assemble commands from class type

* remove unused counter from SamplePulse

* update unittest

* refactor schedule-related interface

* unify exceptions into PulseError

* add a test

* add decorator

* update assembler

* update lo frequency

* add check length of mem_slots

* remove all changes except for channels

* remove all changes except for commands-rework

* lint

* make mem_slots independent of qubits

* fix a bug in Acquire constructor with default values

* refactor meas_opts

* make sample pulses hashable

* make commands callable

* make channels hashable

* add working example code of Schedule

* reorder class defs and add prints

* simplify Schedule

* rename PulseChannel->Channel

* lint

* add getters for qobj creation

* simplify snapshot command

* rename Pulse -> Instruction

* simplify Instruction

* refactor schedule-related interface

* unify exceptions into PulseError

* add check length of mem_slots

* remove all changes except for commands-rework

* make sample pulses hashable

* fix a bug in Acquire constructor with default values

* refactor meas_opts

* fix lint

mtreinish committed:
Test to verify mpl draws an empty circuit empty (#2105)

* Test to verify mpl draws an empty circuit empty

This commit adds a test that verifies that when we call the mpl backend
it draws an empty image of the expected size. This is just a white image
background of the default size.

Related to #1845 and #2097

* Skip if matplotlib is not installed

* Only import from matplotlib if installed

1ucian0 committed:
"swap followed by measure" optimization pass (#1890)

1ucian0 committed:
Performance issue because optimizations-until-fixed-point approach (#2107)

1ucian0 committed:
swap_mapper as a pass (#2094)

* qiskit/mapper/mapping.py moved to qiskit/transpiler/passes/mapping/legacy_swap.py

* LegacySwap added to the mapper meta tests

* layer_permutation params

* swap_mapper_layer_update params

* circuit_graph -> dag

* lint

* seed update

* long line

itoko committed:
Merge branch 'openpulse-schedule' of https://github.com/nkanazawa1989/qiskit-terra into openpulse-schedule

itoko committed:
update a test due to change in channels

itoko committed:
rename begin/end_time to start/stop_time

itoko commented:
I understand `TimeslotOccupancy` is weird. How about `TimeslotAllocation` (still weired?) or `TimeslotCollection` (too general?)?

itoko committed:
Merge branch 'master' into openpulse-schedule

taalexander commented:
I am ok with TimeslotCollection/Timeslot or Timeslot/ChannelTimeslot, or ScheduleTimeslot/Timeslot.

taalexander commented:
Looks good to go after the above change.

maddy-tod committed:
Choose how to justify circuit with latex drawer (#1977)

* Fixed #1685 by adding a dictionary to the DAGCircuit class which keeps track of where a node is in the circuit, by mapping the number of a node to its position from the circuit.data object

* Done some further testing and it seems that all that is needed is the lexicographical topological sort - the key doesn't make a difference

* Removed all code relating to the key function

* Updated the tests to reflect the changes to the order the nodes are output in

* Updating based on lint

* Start of work on having the text circuits be left justified - seems to work, more testing needed

* Updated to get around gates that would overlap

* Refactoring

* Changes to respect order gates were declared in

* Fixed a bug where the lines sometimes didn't appear

* Added in parameter to say how you want the circuit to be justified

* Left and None justify working

* Right justify now working

* Fixed bug where reverse bits didn't work

* Fixed a bug where gates were inlined even though they overlapped

* Fixed bug where resets didn't draw

* Updated the testing to reflect the new outputs

* Fixed up lint and comments

* Removed unintentional changes from other files

* Removed an old image file

* Moved the justify code into a method in _utils

* Fixed merge conflict

* Added in more testing

* Updated CHANGELOG

* tidying up

* Updated LaTeX drawer to accept the justify parameter

* removed some unintentional edits to _text

* Fixed merge

* Fixed based on review comments

* Updated test based on #2039

* Removed some files that were moved

* Added files that were removed by merge back

* Removed some changes not relevant to this PR

* Removed an old line from CHANGELOG

* Updated how the image depth is calculated

* Fix linter

* Removed a line that said an error could be raised - no longer the case

kdk committed:
Conditionals must be updated to match schema (#1778)

* Update qobj JSON schema to allow bfuncs.

bfunc's were defined previously, but a higher precedence condition
required `register` and `memory` attributes to be arrays on all QASM
instructions, while bfuncs define them as single integers. This
fix moves the "memory/register should be an array" condition to the
specific QASM instructions which require it.

* Update circuits_to_qobj to emit qobj-compliant conditionals.

itoko committed:
rename TimeslotOccupancy to TimeslotCollection

itoko committed:
remove __lshift__ syntax sugar (will be readd later)

itoko committed:
fix a bug in type hint

ewinston committed:
make representation method of Unitary object more public (#2113)

Also, implement __str__ and __repr__ methods.

godott committed:
Gate cancellation transpiler pass based on commutation relations (#2012)

* Gate cancellation pass and its tests added; commutation_analysis pass updated

* removed print statement in commutation_analysis

* linting

* removed print statement from test

* linting

* remove redundant imports and repeatedly defined tests

* remove print statements

* update transpiler to not need backend for basis and coupling if pm

* clean up gate_cancellation pass

* rename pass to commutative_cancellation. add approximate check to inst compare. clean tests

* more cleanup

* Fixed a bug in commutative_cancellation that cancels T and S gates; Merge T and S gates with Rz gate and U1 gate; Fixed test_gate_commutation.py and renamed to test_commutative_cancellation.py

* lint

* changelog

itoko committed:
Merge branch 'master' into openpulse-schedule
